go/vt/vtgate/planbuilder/operators/horizon_expanding.go

```diff
@@ -0,0 +1,195 @@
+		op = &Limit{
+			Source: op,
+			AST:    sel.Limit,
+		}
+	}
+
+	return op, rewrite.NewTree("expand horizon into smaller components", op), nil
+}
+
+func checkInvalid(aggregations []Aggr, horizon horizonLike) error {
+	for _, aggregation := range aggregations {
+		if aggregation.Distinct {
+			return errHorizonNotPlanned()
+		}
+	}
+	if _, isDerived := horizon.(*Derived); isDerived {
+		return errHorizonNotPlanned()
+	}
+	return nil
+}
+
+func createProjectionFromSelect(ctx *plancontext.PlanningContext, horizon horizonLike) (out ops.Operator, err error) {
+	qp, err := horizon.getQP(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if !qp.NeedsAggregation() {
+		projX, err := createProjectionWithoutAggr(qp, horizon.src())
+		if err != nil {
+			return nil, err
+		}
+		if derived, isDerived := horizon.(*Derived); isDerived {
+			id := derived.TableId
+			projX.TableID = &id
+			projX.Alias = derived.Alias
+		}
+		out = projX
+
+		return out, nil
+	}
+
+	err = checkAggregationSupported(horizon)
+	if err != nil {
+		return nil, err
+	}
+
+	aggregations, err := qp.AggregationExpressions(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := checkInvalid(aggregations, horizon); err != nil {
+		return nil, err
+	}
+
+	a := &Aggregator{
+		Source:       horizon.src(),
+		Original:     true,
+		QP:           qp,
+		Grouping:     qp.GetGrouping(),
+		Aggregations: aggregations,
+	}
+
+	if derived, isDerived := horizon.(*Derived); isDerived {
+		id := derived.TableId
+		a.TableID = &id
+		a.Alias = derived.Alias
+	}
+
+outer:
+	for colIdx, expr := range qp.SelectExprs {
+		ae, err := expr.GetAliasedExpr()
+		if err != nil {
+			return nil, err
+		}
+		addedToCol := false
+		for idx, groupBy := range a.Grouping {
+			if ctx.SemTable.EqualsExprWithDeps(groupBy.SimplifiedExpr, ae.Expr) {
+				if !addedToCol {
+					a.Columns = append(a.Columns, ae)
+					addedToCol = true
+				}
+				if groupBy.ColOffset < 0 {
+					a.Grouping[idx].ColOffset = colIdx
+				}
+			}
+		}
+		if addedToCol {
+			continue
+		}
+		for idx, aggr := range a.Aggregations {
+			if ctx.SemTable.EqualsExprWithDeps(aggr.Original.Expr, ae.Expr) && aggr.ColOffset < 0 {
+				a.Columns = append(a.Columns, ae)
+				a.Aggregations[idx].ColOffset = colIdx
+				continue outer
+			}
+		}
+		return nil, vterrors.VT13001(fmt.Sprintf("Could not find the %s in aggregation in the original query", sqlparser.String(ae)))
+	}
+
+	return a, nil
+}
+
+func createProjectionWithoutAggr(qp *QueryProjection, src ops.Operator) (*Projection, error) {
+	proj := &Projection{
+		Source: src,
+	}
+
+	for _, e := range qp.SelectExprs {
+		if _, isStar := e.Col.(*sqlparser.StarExpr); isStar {
+			return nil, errHorizonNotPlanned()
+		}
+		ae, err := e.GetAliasedExpr()
+
+		if err != nil {
+			return nil, err
+		}
+		expr := ae.Expr
+		if sqlparser.ContainsAggregation(expr) {
+			aggr, ok := expr.(sqlparser.AggrFunc)
+			if !ok {
+				// need to add logic to extract aggregations and pushed them to the top level
```

Comment:
Should we create an issue for it?

Reply:
at the end of all the operator work. `errHorizonNotPlanned()` will  go away, so this will be handled at some point during removal of the method `errHorizonNotPlanned()`